src/React.purs

```diff
@@ -177,191 +153,166 @@ instance intReactRender :: ReactRender Int
-  }
+  , componentWillUnmount :: ComponentWillUnmount eff
+  | r
+  )
 
--- | Create a component specification with a provided state.
-spec :: forall props state render eff.
-  ReactRender render =>
-  state -> Render props state render eff -> ReactSpec props state render eff
-spec state = spec' \_ -> pure state
-
--- | Create a component specification with a get initial state function.
-spec' :: forall props state render eff.
-  ReactRender render =>
-  GetInitialState props state eff ->
-  Render props state render eff ->
-  ReactSpec props state render eff
-spec' getInitialState renderFn =
-  { render: renderFn
-  , displayName: ""
-  , getInitialState: getInitialState
-  , componentWillMount: \_ -> pure unit
-  , componentDidMount: \_ -> pure unit
-  , componentDidCatch: Nothing
-  , componentWillReceiveProps: \_ _ -> pure unit
-  , shouldComponentUpdate: \_ _ _ -> pure true
-  , componentWillUpdate: \_ _ _ -> pure unit
-  , componentDidUpdate: \_ _ _ -> pure unit
-  , componentWillUnmount: \_ -> pure unit
-  }
+type ReactSpecShouldComponentUpdate props state eff =
+  ( shouldComponentUpdate :: ShouldComponentUpdate props state eff
+  )
 
--- | React class for components.
-foreign import data ReactClass :: Type -> Type
+type ReactSpecAll props state render eff =
+  ReactSpecRequired state render eff (ReactSpecOptional props state eff (ReactSpecShouldComponentUpdate props state eff))
 
--- | Read the component props.
-foreign import getProps :: forall props state eff.
-  ReactThis props state ->
-  Eff (props :: ReactProps | eff) props
-
--- | Read the component refs.
-foreign import getRefs :: forall props state access eff.
-  ReactThis props state ->
-  Eff (refs :: ReactRefs (read :: Read | access) | eff) Refs
+type ReactSpecPure props state render eff =
+  ReactSpecRequired state render eff (ReactSpecOptional props state eff ())
 
--- | Ref type.  You can store `Ref` types on `Refs` object (which in
--- | corresponds to `this.refs`).  Use `ReactDOM.refToNode` if you want to
--- | store a `DOM.Node.Types.Node`
-foreign import data Ref :: Type
-
-foreign import readRefImpl :: forall props state access eff.
+-- | The signature for a ReactClass constructor. A constructor takes the
+-- | `ReactThis` context and returns a record with appropriate lifecycle
+-- | methods.
+type ReactClassConstructor props state render eff r =
   ReactThis props state ->
+  Eff
+    ( props :: ReactProps
+    , state :: ReactState Disallowed
+    | eff
+    ) (Record (ReactSpecRequired state render eff r))
+
+-- | Creates a `ReactClass`` inherited from `React.Component`.
+component
+  :: forall props state render eff r x
+   . Union (ReactSpecRequired (Record state) render eff r) x (ReactSpecAll (Record props) (Record state) render eff)
+  => RowLacks "children" props
+  => RowLacks "key" props
+  => ReactRender render
+  => String
+  -> ReactClassConstructor (Record props) (Record state) render eff r
```

Comment:
Note that I've required both props and state to be Records, as this is what React expects. I've removed the state boxing completely.